src/templatesDatasource.ts

```diff
@@ -0,0 +1,280 @@
+    };
+
+    return this.backendSrv.datasourceRequest(options).then(res => {
+      this.cachedDomains = res.data.items;
+      return this.cachedDomains;
+    });
+  }
+
+  private request(query: Partial<TemplateQuery>, data: string) {
+    const { domain, domainVersion } = query;
+
+    const options: any = {
+      url: `${this.apiUrl}/cognitetemplatesapi/${this.project}/domains/${domain}/${domainVersion}/graphql`,
+      method: 'POST',
+      data: {
+        query: data,
+      },
+    };
+
+    return this.backendSrv.datasourceRequest(options);
+  }
+
+  private postQuery(query: Partial<TemplateQuery>, payload: string) {
+    return this.request(query, payload)
+      .then((results: any) => {
+        return { query, results };
+      })
+      .catch((err: any) => {
+        if (err.data && err.data.error) {
+          throw {
+            message: `GraphQL error: ${err.data.error.reason}`,
+            error: err.data.error,
+          };
+        }
+
+        throw err;
+      });
+  }
+
+  private createQuery(
+    query: TemplateQuery,
+    range: TimeRange | undefined,
+    scopedVars: ScopedVars | undefined = undefined
+  ) {
+    let payload = query.queryText;
+    if (range) {
+      payload = payload.replace(/\$timeFrom/g, range.from.valueOf().toString());
+      payload = payload.replace(/\$timeTo/g, range.to.valueOf().toString());
+    }
+    payload = this.templateSrv.replace(payload, scopedVars);
+
+    return this.postQuery(query, payload);
+  }
+
+  private static getDocs(resultsData: any, dataPath: string): any[] {
+    if (!resultsData) {
+      throw 'resultsData was null or undefined';
+    }
+    const data = dataPath.split('.').reduce((d: any, p: any) => {
+      if (!d) {
+        return null;
+      }
+      return d[p];
+    }, resultsData.data);
+    if (!data) {
+      const errors: any[] = resultsData.errors;
+      if (errors && errors.length !== 0) {
+        throw errors[0];
+      }
+      throw `Your data path did not exist! dataPath: ${dataPath}`;
+    }
+    if (resultsData.errors) {
+      // There can still be errors even if there is data
+      console.log('Got GraphQL errors:');
+      console.log(resultsData.errors);
+    }
+    const docs: any[] = [];
+    const pushDoc = (originalDoc: object) => {
+      docs.push(flatten(originalDoc));
+    };
+    if (Array.isArray(data)) {
+      for (const element of data) {
+        pushDoc(element);
+      }
+    } else {
+      pushDoc(data);
+    }
+    return docs;
+  }
+  private static getDataPathArray(dataPathString: string): string[] {
+    const dataPathArray: string[] = [];
+    for (const dataPath of dataPathString.split(',')) {
+      const trimmed = dataPath.trim();
+      if (trimmed) {
+        dataPathArray.push(trimmed);
+      }
+    }
+    if (!dataPathArray) {
+      throw 'data path is empty!';
+    }
+    return dataPathArray;
+  }
+
+  async query(options: DataQueryRequest<TemplateQuery>): Promise<DataQueryResponse> {
```

Comment:
Too much nesting here: for -> for -> for -> if -> for -> if

Reply:
Yeah we were unsure if we want to refactor it now, or later. Its copied pasted from a GraphQL Grafana plugin: https://github.com/fifemon/graphql-datasource/blob/master/src/DataSource.ts#L130

Reply:
Also @mortenhillbom since we are copy pasting code, we need to make sure to handle the license correctly. However, we might as well just write this ourself, which will most likely be considered our own work.